src/Compilers/CSharp/Portable/CodeGen/CodeGenerator_RefSafety.cs

```diff
@@ -0,0 +1,121 @@
+
+internal partial class CodeGenerator
+{
+    /// <inheritdoc cref="MightEscapeTemporaryRefs(bool, TypeSymbol, RefKind, ParameterSymbol?, AddressKind?, ImmutableArray{ParameterSymbol})"/>
+    private static bool MightEscapeTemporaryRefs(BoundCall node, bool used, AddressKind? receiverAddressKind)
+    {
+        return MightEscapeTemporaryRefs(
+            used: used,
+            returnType: node.Type,
+            returnRefKind: node.Method.RefKind,
+            thisParameterSymbol: node.Method.TryGetThisParameter(out var thisParameter) ? thisParameter : null,
+            receiverAddressKind: receiverAddressKind,
+            parameters: node.Method.Parameters);
+    }
+
+    /// <inheritdoc cref="MightEscapeTemporaryRefs(bool, TypeSymbol, RefKind, ParameterSymbol?, AddressKind?, ImmutableArray{ParameterSymbol})"/>
+    private static bool MightEscapeTemporaryRefs(BoundObjectCreationExpression node, bool used)
+    {
+        return MightEscapeTemporaryRefs(
+            used: used,
+            returnType: node.Type,
+            returnRefKind: RefKind.None,
+            thisParameterSymbol: null,
+            receiverAddressKind: null,
+            parameters: node.Constructor.Parameters);
+    }
+
+    /// <inheritdoc cref="MightEscapeTemporaryRefs(bool, TypeSymbol, RefKind, ParameterSymbol?, AddressKind?, ImmutableArray{ParameterSymbol})"/>
+    private static bool MightEscapeTemporaryRefs(BoundFunctionPointerInvocation node, bool used)
+    {
+        FunctionPointerMethodSymbol method = node.FunctionPointer.Signature;
+        return MightEscapeTemporaryRefs(
+            used: used,
+            returnType: node.Type,
+            returnRefKind: method.RefKind,
+            thisParameterSymbol: null,
+            receiverAddressKind: null,
+            parameters: method.Parameters);
+    }
+
+    /// <summary>
+    /// Determines whether a 'ref' can be captured by the call (considering only its signature).
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// The emit layer consults this to avoid reusing temporaries that are passed by ref to such methods.
+    /// </para>
+    /// <para>
+    /// This is a heuristic which might have false positives, i.e.,
+    /// it might not recognize that a call cannot capture a 'ref'
+    /// even if the binding-time ref safety analysis would recognize that.
+    /// That is fine, the IL will be correct, albeit less optimal (it might use more temps).
+    /// But we still want some heuristic to avoid regressing IL of common safe cases.
+    /// </para>
+    /// </remarks>
+    private static bool MightEscapeTemporaryRefs(
+        bool used,
+        TypeSymbol returnType,
+        RefKind returnRefKind,
+        ParameterSymbol? thisParameterSymbol,
+        AddressKind? receiverAddressKind,
+        ImmutableArray<ParameterSymbol> parameters)
+    {
+        Debug.Assert(receiverAddressKind is null || thisParameterSymbol is not null);
```

Comment:
It looks like the `receiverAddressKind` is now used only in an assertion, should we eliminate this and the same from the other overloads?